File: frontend/server/tools/update-conversation-topic.js
/**
 * Tool for updating conversation topic
 * Allows agents to set or update the topic title and details for a conversation
 */

import pkg from '@prisma/client';
const { PrismaClient } = pkg;

const prisma = new PrismaClient();

/**
 * Updates the topic for a conversation
 * @param {Object} params - Parameters for updating topic
 * @param {string} params.conversation_id - The conversation ID
 * @param {string} params.topic_title - Concise topic title (max 200 chars)
 * @param {string} params.topic_details - Detailed topic description
 * @returns {Object} Result of the operation
 */
export async function updateConversationTopic({ conversation_id, topic_title, topic_details }) {
  try {
    // Validate inputs
    if (!conversation_id) {
      return { 
        success: false, 
        error: 'conversation_id is required' 
      };
    }

    if (!topic_title || topic_title.trim().length === 0) {
      return { 
        success: false, 
        error: 'topic_title is required and cannot be empty' 
      };
    }

    // Truncate title if too long
    const truncatedTitle = topic_title.substring(0, 200);

    // Update the conversation
    const updated = await prisma.conversations.update({
      where: { id: parseInt(conversation_id) },
      data: {
        topic_title: truncatedTitle,
        topic_details: topic_details || null,
        updated_at: new Date()
      }
    });

    // Emit SSE event if available
    if (global.currentSseEmitter) {
      global.currentSseEmitter('topic_updated', {
        conversation_id: updated.id,
        topic_title: updated.topic_title,
        topic_details: updated.topic_details
      });
    }

    return {
      success: true,
      message: 'Topic updated successfully',
      conversation_id: updated.id,
      topic_title: updated.topic_title,
      topic_details: updated.topic_details
    };

  } catch (error) {
    console.error('Error updating conversation topic:', error);
    return {
      success: false,
      error: error.message
    };
  }
}

// Tool definition for agents
export const updateConversationTopicTool = {
  name: 'update_conversation_topic',
  description: 'Update the topic title and details for the current conversation. Use this to set a clear, concise topic that summarizes what the conversation is about.',
  input_schema: {
    type: 'object',
    properties: {
      conversation_id: {
        type: 'string',
        description: 'The ID of the conversation to update'
      },
      topic_title: {
        type: 'string',
        description: 'A concise topic title (max 200 characters) that summarizes the conversation'
      },
      topic_details: {
        type: 'string',
        description: 'Optional detailed description of the topic, including key context, goals, or important information'
      }
    },
    required: ['conversation_id', 'topic_title']
  }
};

// Export for use in orchestrators
export default updateConversationTopic;

File: frontend/python-tools/update_conversation_topic.py
#!/usr/bin/env python3
"""
Update conversation topic tool for bash agents
"""

import os
import sys
import argparse
import sqlite3
from datetime import datetime

def update_conversation_topic(conversation_id, topic_title, topic_details=None):
    """Update the topic for a conversation"""
    
    # Database path
    db_path = os.path.join(os.path.dirname(__file__), '..', 'prisma', 'dev.db')
    
    try:
        # Connect to database
        conn = sqlite3.connect(db_path)
        cursor = conn.cursor()
        
        # Update the conversation
        if topic_details:
            cursor.execute('''
                UPDATE conversations 
                SET topic_title = ?, topic_details = ?, updated_at = ?
                WHERE id = ?
            ''', (topic_title, topic_details, datetime.now().isoformat(), conversation_id))
        else:
            cursor.execute('''
                UPDATE conversations 
                SET topic_title = ?, updated_at = ?
                WHERE id = ?
            ''', (topic_title, datetime.now().isoformat(), conversation_id))
        
        # Check if update was successful
        if cursor.rowcount == 0:
            print(f"Error: Conversation {conversation_id} not found")
            return False
        
        conn.commit()
        conn.close()
        
        print(f"Successfully updated topic for conversation {conversation_id}")
        print(f"Topic: {topic_title}")
        if topic_details:
            print(f"Details: {topic_details}")
        
        return True
        
    except Exception as e:
        print(f"Error updating conversation topic: {e}")
        return False

def main():
    parser = argparse.ArgumentParser(description='Update conversation topic')
    parser.add_argument('conversation_id', type=int, help='Conversation ID')
    parser.add_argument('topic_title', help='Concise topic title (max 200 chars)')
    parser.add_argument('--details', help='Optional detailed description of the topic')
    
    args = parser.parse_args()
    
    # Get conversation ID from environment if not provided and argument is 0
    if args.conversation_id == 0:
        conv_id = os.environ.get('ESPRESSOBOT_CONVERSATION_ID')
        if conv_id:
            args.conversation_id = int(conv_id)
        else:
            print("Error: No conversation ID provided and ESPRESSOBOT_CONVERSATION_ID not set")
            sys.exit(1)
    
    # Truncate title if too long
    topic_title = args.topic_title[:200]
    
    success = update_conversation_topic(
        args.conversation_id,
        topic_title,
        args.details
    )
    
    sys.exit(0 if success else 1)

if __name__ == '__main__':
    main()